azurerm/internal/services/network/express_route_connection_resource.go

```diff
@@ -0,0 +1,349 @@
+	}
+
+	if v, ok := d.GetOk("routing"); ok {
+		expressRouteConnectionParameters.ExpressRouteConnectionProperties.RoutingConfiguration = expandExpressRouteConnectionRouting(v.([]interface{}))
+	}
+
+	future, err := client.CreateOrUpdate(ctx, id.ResourceGroup, id.ExpressRouteGatewayName, id.Name, expressRouteConnectionParameters)
+	if err != nil {
+		return fmt.Errorf("creating/updating %s: %+v", id, err)
+	}
+
+	if err := future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("waiting for creation/update of %s: %+v", id, err)
+	}
+
+	d.SetId(id.ID())
+
+	return resourceExpressRouteConnectionRead(d, meta)
+}
+
+func resourceExpressRouteConnectionRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Network.ExpressRouteConnectionsClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.ExpressRouteConnectionID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Get(ctx, id.ResourceGroup, id.ExpressRouteGatewayName, id.Name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("retrieving %s: %+v", *id, err)
+	}
+
+	d.Set("name", id.Name)
+	d.Set("express_route_gateway_id", parse.NewExpressRouteGatewayID(id.SubscriptionId, id.ResourceGroup, id.ExpressRouteGatewayName).ID())
+
+	if props := resp.ExpressRouteConnectionProperties; props != nil {
+		d.Set("routing_weight", props.RoutingWeight)
+		d.Set("authorization_key", props.AuthorizationKey)
+		d.Set("enable_internet_security", props.EnableInternetSecurity)
+
+		circuitPeeringID := ""
+		if v := props.ExpressRouteCircuitPeering; v != nil {
+			circuitPeeringID = *v.ID
+		}
+		peeringId, err := parse.ExpressRouteCircuitPeeringID(circuitPeeringID)
+		if err != nil {
+			return err
+		}
+		d.Set("express_route_circuit_peering_id", peeringId.ID())
+
+		routing, err := flattenExpressRouteConnectionRouting(props.RoutingConfiguration)
+		if err != nil {
+			return err
+		}
+		if err := d.Set("routing", routing); err != nil {
+			return fmt.Errorf("setting `routing`: %+v", err)
+		}
+	}
+
+	return nil
+}
+
+func resourceExpressRouteConnectionDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Network.ExpressRouteConnectionsClient
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.ExpressRouteConnectionID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	future, err := client.Delete(ctx, id.ResourceGroup, id.ExpressRouteGatewayName, id.Name)
+	if err != nil {
+		return fmt.Errorf("deleting %s: %+v", *id, err)
+	}
+
+	if err := future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("waiting for deletion of %s: %+v", *id, err)
+	}
+
+	return nil
+}
+
+func expandExpressRouteConnectionRouting(input []interface{}) *network.RoutingConfiguration {
+	if len(input) == 0 {
```

Comment:
We should also nil check here as both fields are optional:

```suggestion
	if len(input) == 0 || input[0] == nil {
```

Reply:
Updated